### PR TITLE
Mediawiki

### PR DESCRIPTION
**Pull request by iaforek**

### CODE REVIEW

File: samples/mediawiki/index.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { ReadableApp } from "@scramjet/types";
+import { PassThrough } from "stream";
+import EventSource = require("eventsource");
+
+const url = 'https://stream.wikimedia.org/v2/stream/recentchange';
+
+const app: ReadableApp<string> = async function (_stream) {
+  const outputStream = new PassThrough({ encoding: 'utf-8' });
```
Comment (MichalCz) [inline samples/mediawiki/index.ts:8]:
```suggestion
  const outputStream = new PassThrough({ objectMode: true });
```

File: samples/mediawiki/index.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  eventSource.onopen = function (event) {
+    console.log('--- Opened connection.');
+  };
+
+  eventSource.onerror = function (event) {
+    console.error('--- Encountered error', event);
+  };
+
+  eventSource.onmessage = function (event) {
+    outputStream.write(JSON.stringify(event.data));
```
Comment (MichalCz) [inline samples/mediawiki/index.ts:21]:
```suggestion
    if (!outputStream.write(JSON.stringify(event.data))) /// do something to pause the reading
```

File: samples/mediawiki/index.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  };
+
+  return outputStream;
```
Comment (MichalCz) [inline samples/mediawiki/index.ts:24]:
```suggestion
  outputStream.on("drain", () => eventSource.resume()); // in essence restart reading

  return outputStream;
```

File: samples/mediawiki/index.ts (hunk 1 above)
Comment (iaforek) [reply samples/mediawiki/index.ts]: Done. Works okay.

File: samples/mediawiki/index.ts (hunk 2 above)
Comment (iaforek) [reply samples/mediawiki/index.ts]: Done.

File: samples/mediawiki/index.ts (hunk 3 above)
Comment (iaforek) [reply samples/mediawiki/index.ts:34]: How do I emit or trigger drain on paused STH stream?

Comment (iaforek) [reply samples/mediawiki/index.ts:40]: nvm `si inst invokeEvent`

Comment (iaforek) [reply samples/mediawiki/index.ts:50]: Done.

**Comment by iaforek** [comment]
@MichalCz Take a look again. I addressed all your comments. You can pause and resume now.

File: samples/mediawiki/index.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import { ReadableApp } from "@scramjet/types";
+import { PassThrough } from "stream";
+import EventSource = require("eventsource");
+
+const url = 'https://stream.wikimedia.org/v2/stream/recentchange';
+
```
Comment (MichalCz) [inline samples/mediawiki/index.ts:6]:
```suggestion
let streaming = false;.
```

File: samples/mediawiki/index.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+function init(outputStream, filter) {
+  let eventSource = new EventSource(url);
```
Comment (MichalCz) [inline samples/mediawiki/index.ts:8]:
```suggestion
  if (streaming) return;
  streaming = true;

  let eventSource = new EventSource(url);
```

File: samples/mediawiki/index.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  eventSource.onmessage = function (event) {
+    const data = [event.data];
+    const result = data.map(data => JSON.parse(data)).filter(filter)[0];
+    if (result) {
+      const isOkayToContinue = outputStream.write(result);
+      if (!isOkayToContinue) {
+        // Wait for drain.
+        console.log('--- Pause and wait for drain');
+        eventSource.close();
```
Comment (MichalCz) [inline samples/mediawiki/index.ts:26]:
```suggestion
        eventSource.close();
        streaming = false;
```

**Review by MichalCz — APPROVED** [review]
Ok, a quick suggestion so that we don't accidentally connect twice and leak a reference there (although I don't think there's much of a chance).

**Pull request merged**